metricflow-semantics/metricflow_semantics/specs/spec_set.py

```diff
@@ -0,0 +1,213 @@
+                self.time_dimension_specs,
+                self.entity_specs,
+                self.group_by_metric_specs,
+                self.metric_specs,
+                self.metadata_specs,
+            )
+        )
+
+    def transform(  # noqa: D102
+        self, transform_function: InstanceSpecSetTransform[TransformOutputT]
+    ) -> TransformOutputT:
+        return transform_function.transform(self)
+
+    @property
+    def metric_time_specs(self) -> Sequence[TimeDimensionSpec]:
+        """Returns any specs referring to metric time at any grain."""
+        return tuple(
+            time_dimension_spec
+            for time_dimension_spec in self.time_dimension_specs
+            if time_dimension_spec.is_metric_time
+        )
+
+    @staticmethod
+    def create_from_specs(specs: Sequence[InstanceSpec]) -> InstanceSpecSet:  # noqa: D102
+        return group_specs_by_type(specs)
+
+
+TransformOutputT = TypeVar("TransformOutputT")
+
+
+class InstanceSpecSetTransform(Generic[TransformOutputT], ABC):
+    """Function to use for transforming spec sets."""
+
+    @abstractmethod
+    def transform(self, spec_set: InstanceSpecSet) -> TransformOutputT:  # noqa: D102
+        pass
+
+
+@dataclass
+class _GroupSpecByTypeVisitor(InstanceSpecVisitor[None]):
+    """Groups a spec by type into an `InstanceSpecSet`."""
+
+    metric_specs: List[MetricSpec] = dataclasses.field(default_factory=list)
+    measure_specs: List[MeasureSpec] = dataclasses.field(default_factory=list)
+    dimension_specs: List[DimensionSpec] = dataclasses.field(default_factory=list)
+    entity_specs: List[EntitySpec] = dataclasses.field(default_factory=list)
+    time_dimension_specs: List[TimeDimensionSpec] = dataclasses.field(default_factory=list)
+    group_by_metric_specs: List[GroupByMetricSpec] = dataclasses.field(default_factory=list)
+    metadata_specs: List[MetadataSpec] = dataclasses.field(default_factory=list)
+
+    @override
+    def visit_measure_spec(self, measure_spec: MeasureSpec) -> None:
+        self.measure_specs.append(measure_spec)
+
+    @override
+    def visit_dimension_spec(self, dimension_spec: DimensionSpec) -> None:
+        self.dimension_specs.append(dimension_spec)
+
+    @override
+    def visit_time_dimension_spec(self, time_dimension_spec: TimeDimensionSpec) -> None:
+        self.time_dimension_specs.append(time_dimension_spec)
+
+    @override
+    def visit_entity_spec(self, entity_spec: EntitySpec) -> None:
+        self.entity_specs.append(entity_spec)
+
+    @override
+    def visit_group_by_metric_spec(self, group_by_metric_spec: GroupByMetricSpec) -> None:
+        self.group_by_metric_specs.append(group_by_metric_spec)
+
+    @override
+    def visit_metric_spec(self, metric_spec: MetricSpec) -> None:
+        self.metric_specs.append(metric_spec)
+
+    @override
+    def visit_metadata_spec(self, metadata_spec: MetadataSpec) -> None:
+        self.metadata_specs.append(metadata_spec)
```

Comment:
The rule I've been following is: no `isinstance` calls unless it's in a equals or comparison method. It's a code-smell, and when it's in the code base, it tends to spread.

Reply:
Since we had had this discussion a few times and to avoid having this thread be the place where we debate, let me just say that I see your points and I'm up for continued discussion on the issue.